src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc

```diff
@@ -0,0 +1,571 @@
+        auto reader = make_reader(
+          ntp, tidp, kafka::offset::min(), kafka::offset::max(), 10_KiB, true);
+        auto result = read_all(reader.get());
+        EXPECT_LT(result.size(), expected.size());
+        EXPECT_TRUE(reader->is_end_of_stream());
+    }
+}
+
+TEST_F(l1_reader_test, out_of_range) {
+    auto [ntp, tidp] = make_ntidp("test_topic");
+
+    // No objects, so the whole offset range is out of range.
+    auto reader = make_reader(ntp, tidp);
+
+    auto result = read_all(reader.get());
+
+    EXPECT_TRUE(result.empty());
+    EXPECT_TRUE(reader->is_end_of_stream());
+}
+
+TEST_F(l1_reader_test, missing_object) {
+    auto [ntp, tidp] = make_ntidp("test_topic");
+
+    // Register object in metastore but don't upload.
+    // This would be really bad but it shouldn't crash readers.
+    auto builder = _metastore.object_builder();
+    auto oid = builder->get_or_create_object_for(tidp);
+    builder
+      ->add(
+        oid,
+        l1::metastore::object_metadata::ntp_metadata{
+          .tidp = tidp,
+          .base_offset = kafka::offset{0},
+          .last_offset = kafka::offset{9},
+          .max_timestamp = model::timestamp::now(),
+          .pos = 0,
+          .size = 1000,
+        })
+      .value();
+    builder->finish(oid, 1001, 1500).value();
+
+    l1::metastore::term_offset_map_t term_map;
+    term_map[tidp].push_back(
+      l1::metastore::term_offset{
+        .term = model::term_id{1},
+        .first_offset = kafka::offset{0},
+      });
+
+    _metastore.add_objects(std::move(builder), term_map).get().value();
+
+    auto reader = make_reader(ntp, tidp);
+    auto result = read_all(reader.get());
+    EXPECT_TRUE(result.empty());
+    EXPECT_TRUE(reader->is_end_of_stream());
+}
+
+TEST_F(l1_reader_test, empty_offset_range) {
+    auto [ntp, tidp] = make_ntidp("test_topic");
+
+    // Write some initial objects.
+    auto batches
+      = model::test::make_random_batches(model::offset{0}, 250).get();
+
+    for (auto start = 0; start <= 200; start += 50) {
+        std::vector<tidp_batches_t> tidp_batches;
+        auto subbatches = slice(batches, start, 50);
+        tidp_batches.emplace_back(tidp, std::move(subbatches));
+        make_l1_objects(tidp_batches);
+    }
+
+    // Mimic an object from which all partition data has been compacted.
+    // The object has no data for the partition, but is registered
+    // to cover a non-empty offset range in the metastore.
+    auto meta_builder = _metastore.object_builder();
+
+    auto oid = meta_builder->get_or_create_object_for(tidp);
+
+    auto buf = iobuf{};
+    auto builder = l1::object_builder::create(
+      make_iobuf_ref_output_stream(buf), {});
+
+    auto obj_info = builder->finish().get();
+    builder->close().get();
+
+    _io.put_object(oid, std::move(buf));
+
+    auto missing = kafka::next_offset(
+      model::offset_cast(batches.back().last_offset()));
```

Comment:
nit: `missing` is a bit confusing to me because offset -1 also means "missing" in the kafka API. Can we call this high_watermark or something a bit more standard in kafka terms?